File: scripts/plot_damage_summary.py
#!/usr/bin/env python3
"""
plot_damage_summary.py

Summary biplot of aDNA damage across all taxa from workflow damage TSV outputs.

One point per taxon:
  X axis : baseline classified k-mer rate  (1 - plateau_frac_unclassified)
  Y axis : damage rate (%)                 (damage_score × 100; negative scores shown as 0)
  Size   : log10(n_reads)  — larger = more reads
  Color  : -log10(p-value) — brighter = more significant
  Shape  : △ (significant, p < threshold)  ○ (not significant)

Usage:
  python plot_damage_summary.py \\
    --stats  sample_stats.tsv \\
    --output damage_summary.pdf \\
    [--pvalue-threshold 0.05] \\
    [--min-reads 10]
"""

import argparse
import sys

import matplotlib
matplotlib.use("Agg")
import matplotlib.pyplot as plt
import matplotlib.colors as mcolors
import matplotlib.cm as cm
from matplotlib.lines import Line2D
import numpy as np
import pandas as pd

from hit_species_selection import select_hit_species


# marker size: area = (log10(n) * SIZE_SCALE) ** SIZE_EXP
SIZE_SCALE = 8.0
SIZE_EXP   = 1.5


def marker_area(log10_n: np.ndarray) -> np.ndarray:
    return (np.maximum(log10_n, 0) * SIZE_SCALE) ** SIZE_EXP


def load_stats(path: str, pvalue_threshold: float, min_reads: int) -> pd.DataFrame:
    df = pd.read_csv(path, sep="\t")
    df = df[df["end"] == "5prime"].copy()
    df = df[df["n_reads"] >= min_reads].copy()

    df["x"] = 1.0 - df["plateau_frac_unc"]
    # clamp negative damage rates to 0
    df["y"] = df["damage_score"].clip(lower=0) * 100.0
    df["neg_log10_p"] = -np.log10(df["damage_pvalue"].clip(lower=1e-300))
    df["log10_n"] = np.log10(df["n_reads"].clip(lower=1))
    df["significant"] = (
        df["damage_pvalue"].lt(pvalue_threshold) & df["damage_score"].gt(0)
    )

    # label: prefer species_name, fall back to taxid string
    if "species_name" in df.columns:
        df["label"] = df["species_name"].fillna("").astype(str)
        taxid_mask = df["label"] == ""
        if taxid_mask.any() and "taxid" in df.columns:
            df.loc[taxid_mask, "label"] = df.loc[taxid_mask, "taxid"].astype(str)
    elif "taxid" in df.columns:
        df["label"] = df["taxid"].astype(str)
    else:
        df["label"] = ""

    return df


def make_plot(
    df: pd.DataFrame,
    pvalue_threshold: float,
    output_path: str,
    max_pos_x: float | None,
    hit_species: set[str] | None = None,
):
    vmax = max(float(df["neg_log10_p"].dropna().max()), 2.0)
    norm = mcolors.Normalize(vmin=0, vmax=vmax)
    cmap = matplotlib.colormaps["viridis"]

    fig, ax = plt.subplots(figsize=(9, 6.5))

    for sig, marker in [(True, "^"), (False, "o")]:
        sub = df[df["significant"] == sig]
        if sub.empty:
            continue
        colors = cmap(norm(sub["neg_log10_p"].values))
        sizes  = marker_area(sub["log10_n"].values)
        ax.scatter(
            sub["x"], sub["y"],
            c=colors,
            s=sizes,
            marker=marker,
            edgecolors="white",
            linewidths=0.5,
            alpha=0.75,
            zorder=3,
        )

    # reference line
    ax.axhline(0, color="gray", lw=0.8, ls="--", alpha=0.5, zorder=1)

    # annotate only selected hit species from --hits filtering
    if hit_species is None:
        hit_species = set()
    to_annotate = df[df["label"].isin(hit_species)].copy()
    for _, row in to_annotate.iterrows():
        label = row["label"]
        ax.annotate(
            label,
            xy=(row["x"], row["y"]),
            xytext=(4, 4),
            textcoords="offset points",
            fontsize=6.5,
            color="#333333",
            va="bottom",
        )

    # colorbar
    sm = cm.ScalarMappable(cmap=cmap, norm=norm)
    sm.set_array([])
    cb = fig.colorbar(sm, ax=ax, pad=0.02, fraction=0.03)
    cb.set_label("-log\u2081\u2080(p-value)", fontsize=8)
    cb.ax.tick_params(labelsize=7)

    # legend — shapes + representative sizes
    legend_handles = [
        Line2D([0], [0], marker="^", color="none",
               markerfacecolor="gray", markeredgecolor="white",
               markeredgewidth=0.5, markersize=7,
               label=f"p < {pvalue_threshold} (significant)"),
        Line2D([0], [0], marker="o", color="none",
               markerfacecolor="gray", markeredgecolor="white",
               markeredgewidth=0.5, markersize=7,
               label=f"p \u2265 {pvalue_threshold}"),
    ]
    for n_ref in [100, 1_000, 10_000, 100_000]:
        s = marker_area(np.array([np.log10(n_ref)]))[0]
        legend_handles.append(
            Line2D([0], [0], marker="o", color="none",
                   markerfacecolor="gray", markeredgecolor="white",
                   markeredgewidth=0.5,
                   markersize=max(np.sqrt(s) / np.pi, 3),
                   label=f"n = {n_ref:,}")
        )
    ax.legend(handles=legend_handles, fontsize=7, frameon=False, loc="upper left")

    ax.set_xlabel("Baseline classified k-mer rate  (1 \u2212 plateau unclassified)", fontsize=9)
    ax.set_ylabel("Damage rate (%)", fontsize=9)
    ax.set_title("aDNA damage summary", fontsize=10)
    ax.set_ylim(bottom=0)
    if max_pos_x is not None:
        ax.set_xlim(right=max_pos_x)
    ax.spines[["top", "right"]].set_visible(False)
    ax.tick_params(labelsize=8)

    fig.tight_layout()
    fig.savefig(output_path, bbox_inches="tight", dpi=150)
    print(f"Saved {output_path}", file=sys.stderr)
    plt.close(fig)


def parse_args():
    p = argparse.ArgumentParser(
        description=__doc__,
        formatter_class=argparse.RawDescriptionHelpFormatter,
    )
    p.add_argument("--stats", required=True,
                   help="*_damage_stats.tsv from aggregate_sample.py")
    p.add_argument("--output", required=True,
                   help="Output plot path (.pdf or .png)")
    p.add_argument("--pvalue-threshold", type=float, default=0.05,
                   help="P-value cutoff for significant shape (default: 0.05)")
    p.add_argument("--min-reads", type=int, default=10,
                   help="Minimum reads to include a taxon (default: 10)")
    p.add_argument("--hits",      default=None,
                   help="Integrated summary table (.tsv/.tsv.gz); hit species are always annotated")
    p.add_argument("--sample-id", default=None,
                   help="Sample ID to filter from --hits table")
    p.add_argument("--hits-required-flags", nargs="+", default=[],
                   help="Required tokens in hit_criteria_flag for --hits selection "
                        "(all must be present)")
    p.add_argument("--max-keys", type=int, default=200,
                   help="Maximum number of hit species to annotate (default: 200; <=0 disables cap)")
    p.add_argument("--max-x", type=float, default=None, dest="max_pos_x",
                   help="Clip x-axis maximum (default: auto)")
    return p.parse_args()


def write_empty_plot(output_path: str, note: str) -> None:
    fig, ax = plt.subplots(figsize=(9, 4.5))
    ax.axis("off")
    ax.text(
        0.5, 0.62,
        "No damage-summary points available",
        ha="center", va="center", fontsize=13, fontweight="bold",
    )
    ax.text(
        0.5, 0.42,
        note,
        ha="center", va="center", fontsize=9.5,
    )
    fig.tight_layout()
    fig.savefig(output_path, bbox_inches="tight", dpi=150)
    plt.close(fig)
    print(f"Saved {output_path} (empty summary)", file=sys.stderr)


def _resolve_required_hit_flags(args: argparse.Namespace) -> list[str]:
    if args.hits_required_flags:
        return args.hits_required_flags
    return ["damage_pvalue", "within_genus_relative_abundance", "classified_rate"]


def main():
    args = parse_args()
    df = load_stats(args.stats, args.pvalue_threshold, args.min_reads)
    if df.empty:
        write_empty_plot(
            args.output,
            "No taxa passed plotting filters (end=5prime and min-reads threshold).",
        )
        sys.exit(0)
    hit_species: set[str] = set()
    required_hit_flags = _resolve_required_hit_flags(args)
    if args.hits and args.sample_id:
        try:
            selected, hit_info = select_hit_species(
                hits_path=args.hits,
                sample_id=str(args.sample_id),
                required_flag_tokens=required_hit_flags,
                max_keys=args.max_keys,
            )
            hit_species = set(selected)
            if hit_info["n_truncated"] > 0:
                print(
                    f"WARNING: selected hit species truncated to {len(hit_species)} "
                    f"(dropped {hit_info['n_truncated']} by --max-keys).",
                    file=sys.stderr,
                )
        except (FileNotFoundError, pd.errors.EmptyDataError, pd.errors.ParserError, EOFError, OSError) as e:
            print(f"WARNING: could not read --hits file ({args.hits}): {e}", file=sys.stderr)
            pass

    make_plot(
        df,
        pvalue_threshold=args.pvalue_threshold,
        output_path=args.output,
        max_pos_x=args.max_pos_x,
        hit_species=hit_species,
    )


if __name__ == "__main__":
    main()

File: scripts/plot_damage_summary.R
#!/usr/bin/env Rscript
# ---------------------------------------------------------------------------
# plot_damage_summary.R
#
# Summary biplot of aDNA damage across all taxa from workflow damage TSVs.
#
# One point per taxon:
#   X axis : baseline classified k-mer rate  (1 - plateau_frac_unc)
#   Y axis : damage rate (%)  (damage_score * 100; negative scores clamped to 0)
#   Size   : log10(n_reads)
#   Fill   : --color-by evenness (default) - evenness_index on mako reversed,
#            so dark = high evenness; or --color-by pvalue for -log10(p)
#   Shape  : filled down-triangle = damage-significant, circle = not
#   Outline: black = all four hit criteria passed, none otherwise
#   Labels : ggrepel with leader segments
#
# Colouring by evenness adds information the rest of the plot does not carry:
# both axes and the marker shape are damage-derived, whereas evenness separates
# genome-wide coverage from reads piled onto a conserved region. Evenness is
# read from the per-sample coverage.tsv, which covers every plotted taxon; the
# integrated summary only retains taxa with complete evidence of all four kinds
# and would leave a large fraction uncoloured.
# ---------------------------------------------------------------------------

suppressPackageStartupMessages({
  library(ggplot2)
  library(readr)
  library(dplyr)
  library(ggrepel)
  library(scales)
})

script_dir <- function() {
  a <- commandArgs(trailingOnly = FALSE)
  f <- sub("^--file=", "", a[grepl("^--file=", a)])
  if (length(f)) dirname(normalizePath(f)) else "."
}
source(file.path(script_dir(), "plot_common.R"))

ALL_HIT_FLAG_TOKENS <- c(
  "damage_pvalue", "evenness_index",
  "within_genus_relative_abundance", "classified_rate"
)

SPEC <- list(
  stats               = list(type = "character", nargs = 1,   default = NULL),
  output              = list(type = "character", nargs = 1,   default = NULL),
  coverage            = list(type = "character", nargs = 1,   default = NULL),
  hits                = list(type = "character", nargs = 1,   default = NULL),
  sample_id           = list(type = "character", nargs = 1,   default = NULL),
  hits_required_flags = list(type = "character", nargs = "+", default = character(0)),
  pvalue_threshold    = list(type = "double",    nargs = 1,   default = 0.05),
  min_reads           = list(type = "integer",   nargs = 1,   default = 10L),
  max_keys            = list(type = "integer",   nargs = 1,   default = 200L),
  max_x               = list(type = "double",    nargs = 1,   default = NULL),
  color_by            = list(type = "character", nargs = 1,   default = "evenness"),
  no_label_repel      = list(type = "flag",      nargs = 1,   default = FALSE)
)

args <- parse_cli(commandArgs(trailingOnly = TRUE), SPEC)
stopifnot(!is.null(args$stats), !is.null(args$output))

# ---------------------------------------------------------------------------
# Load
# ---------------------------------------------------------------------------

stats <- suppressWarnings(readr::read_tsv(args$stats, col_types = readr::cols(
  .default = readr::col_guess(),
  species_name = readr::col_character(),
  end = readr::col_character()
), progress = FALSE))

df <- stats %>%
  filter(.data$end == "5prime", .data$n_reads >= args$min_reads) %>%
  mutate(
    x           = 1 - .data$plateau_frac_unc,
    y           = pmax(.data$damage_score, 0) * 100,
    neg_log10_p = -log10(pmax(.data$damage_pvalue, 1e-300)),
    log10_n     = log10(pmax(.data$n_reads, 1)),
    significant = .data$damage_pvalue < args$pvalue_threshold & .data$damage_score > 0,
    label       = dplyr::coalesce(as.character(.data$species_name), "")
  )
if ("taxid" %in% names(df)) {
  df$label <- ifelse(nzchar(df$label), df$label, as.character(df$taxid))
}

if (nrow(df) == 0L) {
  write_empty_plot(
    args$output,
    "No damage-summary points available",
    "No taxa passed plotting filters (end=5prime and min-reads threshold)."
  )
  quit(save = "no", status = 0)
}

# evenness for colour: prefer per-sample coverage.tsv, fall back to the summary
load_evenness <- function(path, sample_id) {
  if (is.null(path) || is.null(sample_id)) return(NULL)
  d <- tryCatch(
    suppressWarnings(readr::read_tsv(path, col_types = readr::cols(
      .default = readr::col_guess(), sample_id = readr::col_character()
    ), progress = FALSE)),
    error = function(e) {
      message("WARNING: could not read evenness from ", path, ": ", conditionMessage(e))
      NULL
    }
  )
  if (is.null(d) || !"evenness_index" %in% names(d)) return(NULL)
  if ("sample_id" %in% names(d)) {
    d <- d[as.character(d$sample_id) == as.character(sample_id), , drop = FALSE]
  }
  if ("rank" %in% names(d)) {
    d <- d[tolower(trimws(d$rank)) == "species", , drop = FALSE]
  }
  name_col <- if ("species_name" %in% names(d)) "species_name" else "tax_name"
  if (!name_col %in% names(d)) return(NULL)
  d <- d[!is.na(d[[name_col]]), , drop = FALSE]
  d <- d[!duplicated(d[[name_col]]), , drop = FALSE]
  setNames(as.numeric(d$evenness_index), d[[name_col]])
}

if (identical(args$color_by, "evenness")) {
  ev <- load_evenness(if (!is.null(args$coverage)) args$coverage else args$hits,
                      args$sample_id)
  df$evenness_index <- if (is.null(ev)) NA_real_ else unname(ev[df$label])
  n_missing <- sum(is.na(df$evenness_index))
  if (n_missing > 0L) {
    message("NOTE: ", n_missing, " of ", nrow(df),
            " taxa have no evenness_index (drawn grey)")
  }
}

# annotation set and full-hit set
required_flags <- resolve_required_hit_flags(args$hits_required_flags)
sel <- select_hit_species(args$hits, args$sample_id, required_flags, args$max_keys)
if (sel$n_truncated > 0L) {
  message("WARNING: selected hit species truncated to ", length(sel$species),
          " (dropped ", sel$n_truncated, " by --max-keys).")
}
full <- select_hit_species(args$hits, args$sample_id, ALL_HIT_FLAG_TOKENS, 0L)

df$is_full <- df$label %in% full$species
to_label <- df[df$label %in% sel$species, , drop = FALSE]

# ---------------------------------------------------------------------------
# Plot
# ---------------------------------------------------------------------------

if (identical(args$color_by, "evenness")) {
  df$fill_value <- df$evenness_index
  fill_scale <- scale_fill_viridis_c(
    option = "mako", direction = -1, limits = c(0, 1),
    na.value = "#cccccc", name = "evenness_index"
  )
} else {
  df$fill_value <- df$neg_log10_p
  fill_scale <- scale_fill_viridis_c(
    option = "mako", direction = -1,
    na.value = "#cccccc", name = expression(-log[10](p))
  )
}

# shapes 21/25 take a separate fill and outline colour, so the evenness scale
# and the full-hit outline can be encoded independently
df$shape <- ifelse(df$significant, 25L, 21L)

size_breaks <- c(100, 1e3, 1e4, 1e5)

p <- ggplot(df, aes(x = .data$x, y = .data$y)) +
  geom_hline(yintercept = 0, linetype = "dashed", colour = "grey60",
             linewidth = 0.3) +
  geom_point(
    aes(fill = .data$fill_value, size = .data$log10_n,
        shape = .data$shape, colour = .data$is_full),
    alpha = 0.8, stroke = 0.35
  ) +
  fill_scale +
  scale_shape_identity() +
  # "transparent", not NA: a row whose colour maps to NA is dropped by ggplot
  # entirely, which would silently plot only the outlined points.
  scale_colour_manual(
    values = c(`TRUE` = "black", `FALSE` = "transparent"),
    na.value = "transparent", guide = "none"
  ) +
  scale_size_continuous(
    name = "reads",
    range = c(0.6, 7),
    breaks = log10(size_breaks),
    labels = scales::comma(size_breaks)
  ) +
  labs(
    x = "Baseline classified k-mer rate  (1 − plateau unclassified)",
    y = "Damage rate (%)",
    title = "aDNA damage summary"
  ) +
  theme_screen()

if (!is.null(args$max_x)) p <- p + coord_cartesian(xlim = c(NA, args$max_x))

if (nrow(to_label) > 0L) {
  p <- if (isTRUE(args$no_label_repel)) {
    p + geom_text(data = to_label, aes(label = .data$label),
                  size = 2.3, colour = "#333333", hjust = 0, vjust = 0,
                  nudge_x = 0.004, nudge_y = 0.4)
  } else {
    p + geom_text_repel(
      data = to_label, aes(label = .data$label),
      size = 2.3, colour = "#333333",
      segment.colour = "grey55", segment.size = 0.25,
      min.segment.length = 0, box.padding = 0.35, point.padding = 0.2,
      max.overlaps = Inf, seed = 42
    )
  }
}

ggsave(args$output, p, width = 9, height = 6.5, device = pdf_device())
message("Saved ", args$output)
